### PR TITLE
fix some post-processing issues

### LINKED ISSUE #5721: UI only shows if in last pass
## Bevy version

f1be89d45809c71c7d3f9f341f99e80780d69b20

## Relevant system information

Tested on MacOS and iOS, with 4x MSAA

## What you did

I used `UiCameraConfig` to enable `show_ui` on only one pass that was not the last pass (using camera priority.)

## What went wrong

The UI did not show, but I could see it was being rendered using the Metal debugger.

## Additional information

I suspected that the wrong color attachment was being used, so that the rendered UI was accidentally discarded.

So I made the following changes:

```diff
diff --git a/crates/bevy_ui/src/render/pipeline.rs b/crates/bevy_ui/src/render/pipeline.rs
index e199cef52..7914d3b8f 100644
--- a/crates/bevy_ui/src/render/pipeline.rs
+++ b/crates/bevy_ui/src/render/pipeline.rs
@@ -105,7 +105,7 @@ impl SpecializedRenderPipeline for UiPipeline {
             },
             depth_stencil: None,
             multisample: MultisampleState {
-                count: 1,
+                count: 4,
                 mask: !0,
                 alpha_to_coverage_enabled: false,
             },
diff --git a/crates/bevy_ui/src/render/render_pass.rs b/crates/bevy_ui/src/render/render_pass.rs
index 66445c92a..9308b7a26 100644
--- a/crates/bevy_ui/src/render/render_pass.rs
+++ b/crates/bevy_ui/src/render/render_pass.rs
@@ -81,14 +81,10 @@ impl Node for UiPassNode {
         };
         let pass_descriptor = RenderPassDescriptor {
             label: Some("ui_pass"),
-            color_attachments: &[Some(RenderPassColorAttachment {
-                view: &target.view,
-                resolve_target: None,
-                ops: Operations {
-                    load: LoadOp::Load,
-                    store: true,
-                },
-            })],
+            color_attachments: &[Some(target.get_color_attachment(Operations {
+                load: LoadOp::Load,
+                store: true,
+            }))],
             depth_stencil_attachment: None,
         };
```

and it works as expected now, but it's probably not necessary to use MSAA for the UI, so I'm not sure what the correct way to do this is. Obviously hardcoding the multisample count is a bad idea; it would have to be passed through the key if this is actually the correct solution.

### LINKED ISSUE #7435: Post processing doesn't work correctly when using multiple viewports
## Bevy version

0.9 and main

## What you did

In the split_screen example: 
- Enable hdr on both cameras (enabling hdr only on one also breaks bevy, but this is a separate issue).
- Add `BloomSettings::default()` to the left camera
- Use an emissive color for the plane

image

I would have expected the left viewport to have bloom applied to it.

## What went wrong

The bloom effect does not appear in the left viewport.

As you can see here, the bloom effect is indeed computed and generated correctly, but once it gets to the final upscaling pass, the texture modified by the post process is not used as an input.
image

image

I haven't investigated more than that, but it seems like the upscaling node is using the wrong texture when using multiple viewports combined with post processing.

### PR DESCRIPTION
**Pull request by robtfm**
# Objective

fix a few post-processing issues

## Solution

- fxaa now respects viewport
- tonemapping is only done on the final camera for a given target
- upscaling writes to the input texture (if required) instead of the output texture if the active camera is not the final camera for a given target (stops post-processing and UI from being dropped for cameras that are not last with MSAA targets, or the last effect from being dropped for non-MSAA targets)

should fix #7435 and hopefully fix #5721 (haven't tested that)

### CODE REVIEW

**Comment by Keelar** [comment]
I just tested it and can confirm that this does seem to fix #5721

**Comment by robtfm** [comment]
i think there's still one issue remaining: if MSAA is enabled, a first camera renders with an odd number of post-process swaps, and then a second camera runs without doing an msaa resolve (which the UI camera doesn't currently, for example), then the last post-process step will be dropped.

in the current core pipeline this will only occur if you're using MSAA and FXAA at the same time, so I don't think it's a big problem.

i think making MSAA a node in its own right instead of part of the 2/d3d main pass would be the best way to fix this. this is useful anyway as some post effects may want to be pre-msaa resolve. but i'll do that in a separate PR.

File: crates/bevy_render/src/view/mod.rs [hunk 1]
```diff
@@ -218,12 +218,27 @@ impl ViewTarget {
...
+    }
+
     /// The format of the final texture this view will render to
     #[inline]
     pub fn out_texture_format(&self) -> TextureFormat {
         self.out_texture_format
     }
 
+    /// the format of the initial texture that will be written to.
+    pub fn base_texture_format(&self) -> TextureFormat {
```
Comment (IceSentry) [inline crates/bevy_render/src/view/mod.rs:238]: Why not just use `main_texture_format()`? Do you think these might be different in the future?

**Review by IceSentry — APPROVED** [review]
I tested it and can confirm that this fixes the original issue. I haven't tested with an odd number of viewports though.

Code makes sense to me, could extract the `final_order` to a shared fn, but it's not super important.

**Comment by robtfm** [comment]
i just realised why i said on #7435 we needed to keep tonemapping on intermediate cameras and reverse-tonemap in the next camera. this pr would cause fxaa to run before tonemapping if it is not on the last camera...

**Comment by robtfm** [comment]
closing in favour of #7474

**Pull request closed without merging**